proj/hps_accel/src/software_cfu.h

```diff
@@ -0,0 +1,190 @@
+  }
+  return macc_out;
+}
+
+// Very simple storage class.
+// TODO: wrap around at calculated point instead of at end of data
+// TODO: allow input data to be double-buffered
+#define MAX_STORAGE_WORDS 16384
+class Storage {
+ public:
+  void Reset(size_t num_words) {
+    len = num_words;
+    index = 0;
+  }
+  void Store(uint32_t value) { data[index++] = value; }
+  uint32_t Get(size_t n) {
+    if (index >= len) {
+      index = 0;
+    }
+    uint32_t result = data[index + n];
+    if (n == 3) {
+      index += 4;
+    }
+    return result;
+  }
+
+ private:
+  uint32_t data[MAX_STORAGE_WORDS];
+  size_t len;
+  size_t index;
+};
+
+extern Storage filter_storage;
+extern Storage input_storage;
+
+inline uint32_t SetRegister(int funct7, uint32_t rs1, uint32_t rs2) {
+  switch (funct7) {
+    case REG_RESET:
+      filter_storage.Reset(0);
+      return 0;
+    case REG_FILTER_NUM_WORDS:
+      filter_storage.Reset(rs1);
+      return 0;
+    case REG_INPUT_NUM_WORDS:
+      input_storage.Reset(rs1);
+      return 0;
+    case REG_INPUT_OFFSET:
+      macc_input_offset = rs1;
+      return 0;
+    case REG_SET_FILTER:
+      filter_storage.Store(rs1);
+      return 0;
+    case REG_SET_INPUT:
+      input_storage.Store(rs1);
+      return 0;
+    case REG_MACC_INPUT_0:
+      SetMaccInput(0, rs1);
+      return 0;
+    case REG_MACC_INPUT_1:
+      SetMaccInput(1, rs1);
+      return 0;
+    case REG_MACC_INPUT_2:
+      SetMaccInput(2, rs1);
+      return 0;
+    case REG_MACC_INPUT_3:
+      SetMaccInput(3, rs1);
+      return 0;
+    case REG_MACC_FILTER_0:
+      SetMaccFilter(0, rs1);
+      return 0;
+    case REG_MACC_FILTER_1:
+      SetMaccFilter(1, rs1);
+      return 0;
+    case REG_MACC_FILTER_2:
+      SetMaccFilter(2, rs1);
+      return 0;
+    case REG_MACC_FILTER_3:
+      SetMaccFilter(3, rs1);
+      return 0;
+
+    default:
+      printf("\nInvalid SetRegister number %d\n", funct7);
+      return 0;
+  }
+  return 0;
+}
+
+inline uint32_t GetRegister(int funct7, uint32_t rs1, uint32_t rs2) {
+  switch (funct7) {
+    case REG_FILTER_0:
+      return filter_storage.Get(0);
+    case REG_FILTER_1:
+      return filter_storage.Get(1);
+    case REG_FILTER_2:
+      return filter_storage.Get(2);
+    case REG_FILTER_3:
+      return filter_storage.Get(3);
+    case REG_INPUT_0:
+      return input_storage.Get(0);
+    case REG_INPUT_1:
+      return input_storage.Get(1);
+    case REG_INPUT_2:
+      return input_storage.Get(2);
+    case REG_INPUT_3:
+      return input_storage.Get(3);
+    case REG_MACC_OUT:
+      return Macc();
+    default:
+      printf("\nInvalid GetRegister number %d\n", funct7);
+      return 0;
+  }
+}
+
+extern uint32_t ping_storage;
+
+inline uint32_t Ping(uint32_t rs1, uint32_t rs2) {
+  uint32_t result = ping_storage;
+  ping_storage = rs1 + rs2;
+  return result;
+}
+
+};  // namespace soft_cfu
+
+inline uint32_t software_cfu(int funct3, int funct7, uint32_t rs1,
```

Comment:
Hmm maybe we should try and fix LTO in GCC, it might let us avoid this kind of "dump everything in the header and `inline` it" optimisation...